### PR TITLE
 [3/N] Fix cppcoreguidelines-special-member-functions warnings

### PR DESCRIPTION
**Pull request by cyyever**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: aten/src/ATen/TensorIterator.h [hunk 1]
```diff
@@ -250,7 +250,6 @@ struct TORCH_API TensorIteratorBase : public impl::MetaBase {
   using PtrVector = SmallVector<char*, 4>;
   using StrideVector = SmallVector<int64_t, 6>;
 
-  TensorIteratorBase();
```
Comment (Skylion007) [inline aten/src/ATen/TensorIterator.h:253]: This is still default constructable, right?

File: aten/src/ATen/core/Array.h [hunk 2]
```diff
@@ -23,10 +23,16 @@ struct Array {
   C10_HOST_DEVICE Array() = default;
   C10_HOST_DEVICE Array(const Array&) = default;
   C10_HOST_DEVICE Array& operator=(const Array&) = default;
+  C10_HOST_DEVICE Array(Array&&) = default;
```
Comment (Skylion007) [inline aten/src/ATen/core/Array.h:26]: They shouldn't be noexcept?

File: aten/src/ATen/TensorIterator.h (hunk 1 above)
Comment (cyyever) [reply aten/src/ATen/TensorIterator.h:253]: Right, but it's not need to write explicitly.

File: aten/src/ATen/core/Array.h (hunk 2 above)
Comment (cyyever) [reply aten/src/ATen/core/Array.h:26]: Can be, but this class should be replaced by std::array once cuda 12.1 is discard.

File: aten/src/ATen/core/ivalue.h [hunk 3]
```diff
@@ -1352,8 +1352,13 @@ struct TORCH_API IValue final {
...
       } as_device;
     } u;
+    static_assert(std::is_trivially_copyable_v<TriviallyCopyablePayload>);
     at::Tensor as_tensor;
     Payload() : u() {}
+    Payload(const Payload&) = delete;
+    Payload(Payload&&) = delete;
+    Payload& operator=(const Payload&) = delete;
+    Payload& operator=(Payload&&) = delete;
     ~Payload() {}
```
[the remark this answers is not in the document]
Comment (cyyever) [reply aten/src/ATen/core/ivalue.h:1362]: Can't use compiler generated because the union has Tensor.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
need rebase

**Comment by wdvr** [comment]
@pytorchmergebot revert -m "reverting since this started failing a windows test"

[GH job link](https://github.com/pytorch/pytorch/actions/runs/11545772185/job/32133934475) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/7cb3cef05f4b1d1b448a82a01420e2a9ed1ccfe0)

**Comment by wdvr** [comment]
@pytorchmergebot revert -m "reverting since this started failing a windows test" -c ignoredsignal

[GH job link](https://github.com/pytorch/pytorch/actions/runs/11545772185/job/32133934475) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/7cb3cef05f4b1d1b448a82a01420e2a9ed1ccfe0)

**Pull request closed without merging**